compiler/rustc_lint/src/empty_iterator.rs

```diff
@@ -0,0 +1,76 @@
+use crate::{EarlyContext, EarlyLintPass, LintContext};
+
+use rustc_ast::{Expr, ExprKind, Lit, LitKind};
+use rustc_errors::fluent;
+
+declare_lint! {
+    /// The `empty_for_iterator` lint checks for empty `for` iterator.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// for i in 10..0 { /* ... */ }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to have a statement that has no effect.
+    pub EMPTY_FOR_ITERATOR,
+    Warn,
+    "empty for iterator"
+}
+
+declare_lint_pass!(EmptyForIterator => [EMPTY_FOR_ITERATOR]);
+
+impl EarlyLintPass for EmptyForIterator {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        match &expr.kind {
+            ExprKind::ForLoop(_, for_expr, _, _) => {
+                let range_span = expr.span.with_hi(for_expr.span.hi());
+                match for_expr.clone().into_inner().kind {
```

Comment:
[`P` implements `Deref`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ptr/struct.P.html#impl-Deref-for-P%3CT%3E), so you can just use the field directly (similarly for other places): 
```suggestion
                match &for_expr.kind {
```